README.md
# envfiles: Simple layered loading of env files

So you've set up your app to configure itself from environment variables,
awesome, well done!

Now all you need to do is to make sure *all of them* are present before
running your app. Good thing we have env files, right? But wait, they can't
be the same all the time, they're slightly different for tests, Docker,
and for local development... that's annoying.

No biggie though, we can just create several env files and load them...
But that would fail in prod because the env files won't be there...
We could load them only if they exist but that's kinda hacky...

Wouldn't it be neat if we could layer env files? Imagine a `base.env`:

```
CACHE_ENABLED=1
DATABASE_HOST=localhost
DATABASE_NAME=myapp
```

A `test.env` that overrides it:

```
# >> base.env
CACHE_ENABLED=0
DATABASE_NAME=myapp_test
```

Note the first line `# >> base.env`, the `# >> ` is an arbitrary prefix, the
`base.env` is a path *relative to the file being read*.

Then point envfiles to the file you want to load and let it resolve the variables.
For example, you could define *one* environment variable, `ENV_FILE` for instance,
containing a relative path to the env file you want to load:

```python
import envfiles

env_vars = envfiles.load_env_files(os.getenv("ENV_FILE"))
assert env_vars == {
    "CACHE_ENABLED": "0",
    "DATABASE_HOST": "localhost",
    "DATABASE_NAME": "myapp_test",
}
```

Note `envfiles` will **not** mess with your `os.environ`, or attempt parsing or
(de)serializing variables. The output is a dictionary of strings to strings,
what you do with it is entirely your business. Typically you would combine it
with `os.environ` and pass it to your app's configuration solution.

For example, if you're using
[`environ-config`](https://github.com/hynek/environ-config) 💚

```python
import environ
import envfiles
from settings import MyConfig  # your environ-config Config class

env_vars = envfiles.load(os.getenv("ENV_FILE"))

env_vars.update(os.environ)  # actual environment variables have preference
config = environ.to_config(MyConfig, environ=env_vars)
```

## Why?

Frustration, mostly. I've had this issue more times than I can count. I've tried
different libraries but none of them supported layering or sort of did but
messed with the `os.environ` in surprising ways.

The idea behind `envfiles` is to declaratively define how the env files are loaded
and let you use whatever configuration solution you want.

Configuration is the first thing your app does, it should be as quick and
straightforward as possible.

## Alternatives

If `envfiles` is not what you were looking for here are some other options you
may want to consider, all of them more mature and featured than `envfiles`:

- [`dotenv`](https://github.com/theskumar/python-dotenv)
- [`environs`](https://github.com/sloria/environs)


CHANGELOG.md
# `envfiles` changelog

## 0.1.1

- Added support for Python 3.9 and 3.10

**Deprecation notice:** 0.1.1 will be the last version to support Python 3.6.

## 0.1.0

- Added support for Python 3.6 and 3.8

## 0.1.0.dev0

- Initial release

tests/test_envfiles.py
from io import StringIO

import pytest

import envfiles


def test_path_does_not_exist():
    with pytest.raises(envfiles.EnvFilesError):
        envfiles.load("path/to/nowhere")


@pytest.mark.parametrize(
    "contents,expected",
    [
        ("FOO=bar", {"FOO": "bar"}),
        ("EMPTY=", {"EMPTY": ""}),
        ('FOO=bar\nBAR="baz"', {"FOO": "bar", "BAR": "baz"}),
        ('FOO=bar\n# comment\nBAR="baz"\n\n', {"FOO": "bar", "BAR": "baz"}),
    ],
)
def test_env_file(mocker, contents, expected):
    mocker.patch("envfiles.envfiles.open", return_value=StringIO(contents))
    mocker.patch("envfiles.envfiles.Path.exists", return_value=True)

    env_vars = envfiles.load("some/path")

    assert env_vars == expected


def test_nested_env_file(mocker):
    contents = "# >> base.env\nBAR=baz"
    base_contents = "FOO=bar"
    mocker.patch(
        "envfiles.envfiles.open",
        side_effect=[StringIO(contents), StringIO(base_contents)],
    )
    mocker.patch("envfiles.envfiles.Path.exists", return_value=True)

    env_vars = envfiles.load("some/path")

    assert env_vars == {"FOO": "bar", "BAR": "baz"}


def test_nested_env_file_overrides_previous_values(mocker):
    contents = "FOO=override-me\n# >> base.env\nBAR=baz"
    base_contents = "FOO=bar"
    mocker.patch(
        "envfiles.envfiles.open",
        side_effect=[StringIO(contents), StringIO(base_contents)],
    )
    mocker.patch("envfiles.envfiles.Path.exists", return_value=True)

    env_vars = envfiles.load("some/path")

    assert env_vars == {"FOO": "bar", "BAR": "baz"}


def test_nested_env_file_fails_if_self_referencial(mocker):
    contents = "# >> local.env\nBAR=baz"
    mocker.patch("envfiles.envfiles.open", return_value=StringIO(contents))
    mocker.patch("envfiles.envfiles.Path.exists", return_value=True)

    with pytest.raises(envfiles.EnvFilesError):
        envfiles.load("local.env")


def test_malformed_line(mocker):
    with pytest.raises(envfiles.EnvFilesError):
        envfiles.envfiles.process_env_line("FOO")


@pytest.mark.integration
def test_layered_env_files():
    env_vars = envfiles.load("tests/assets/test.env")

    assert env_vars == {
        "CACHE_ENABLED": "0",
        "DATABASE_HOST": "localhost",
        "DATABASE_NAME": "myapp_test",
    }


src/envfiles/envfiles.py
import typing

from pathlib import Path


class EnvFilesError(Exception):
    pass


NESTED_ENV_FILE_PREFIX = "# >> "


def load(
    env_file_path: typing.Union[str, Path], nested_prefix: str = NESTED_ENV_FILE_PREFIX
) -> typing.Dict[str, str]:
    """Loads env files from a relative path to the current working directory.

    Env files contain the familiar KEY=VALUE lines, with VALUE optionally
    being quoted, the return dict will strip the quotes and remove any
    trailing whitespace. Note all names and values are returned as strings.

    Env files may also contain comments in the form of
    `# >> PATH_TO_ENV_FILE`, this will signal EnvFiles to load the nested
    path before continuing.

    Nested env paths need not be defined at the top of the file but
    their contents *will override* any previous entries in the containing
    file.
    """
    path = Path.cwd() / Path(env_file_path)
    if not path.exists():
        raise EnvFilesError(f"Path {path} does not exist")

    env_vars = {}
    with open(path, "r") as fd:
        for line in fd.readlines():
            line = line.strip()
            if line.startswith(nested_prefix):
                nested_env_path = line[len(nested_prefix) :]
                if nested_env_path == env_file_path:
                    raise EnvFilesError(
                        f"Self referencing nested path {nested_env_path}"
                    )
                relative_nested_path = path.parent / nested_env_path
                env_vars.update(load(relative_nested_path))
            elif line.startswith("#") or not line:
                continue
            else:
                name, value = process_env_line(line)
                env_vars[name] = value

    return env_vars


def process_env_line(line: str) -> typing.Tuple[str, str]:
    """Returns a 2-tuple of (name, value) for a line in an env file."""
    try:
        name, value = line.strip().split("=", maxsplit=1)
    except ValueError as exc:
        raise EnvFilesError(f"Error processing line {line}") from exc

    return name, process_value(value)


def process_value(value: str) -> str:
    """Returns a processed value for an environment variable."""
    if len(value) > 0 and value[0] == value[-1] == '"':
        return value[1:-1]

    return value


noxfile.py
import nox


nox.options.stop_on_first_error = True
nox.options.reuse_existing_virtualenvs = True

source_files = ("src/envfiles", "tests", "noxfile.py")


@nox.session
def lint(session):
    session.install("-e", ".[dev]")
    session.run("autoflake", "--in-place", "--recursive", *source_files)
    session.run("seed-isort-config", "--application-directories=src/envfiles")
    session.run("isort", "--project=envfiles", *source_files)
    session.run("black", "--target-version=py36", *source_files)


@nox.session
def check(session):
    session.install("-e", ".[dev]")
    session.run("black", "--check", "--diff", "--target-version=py36", *source_files)
    session.run("flake8", *source_files)
    session.run("mypy", "src/envfiles")
    session.run(
        "isort", "--check", "--diff", "--project=envfiles", "--recursive", *source_files
    )


@nox.session(python=["3.6", "3.7", "3.8", "3.9", "3.10"])
def test(session):
    session.install("-e", ".[test]")
    session.run("python", "-m", "pytest", *session.posargs)


@nox.session
def publish_test(session):
    _publish(session, "testpypi")


@nox.session
def publish(session):
    _publish(session, None)


def _publish(session, repository):
    session.install("flit")
    url = "pypi.org"
    if repository == "testpypi":
        url = f"test.{url}"

    confirm = input(f"Are you sure you want to release to {url}? ")
    if confirm.lower().strip() not in ("y", "yes"):
        session.error("Aborting as requested")

    args = ["flit", "publish"]
    if repository is not None:
        args += ["--repository", repository]

    session.run(*args)
